nav2_collision_monitor/src/collision_detector_node.cpp

```diff
@@ -0,0 +1,317 @@
+// Copyright (c) 2022 Samsung R&D Institute Russia
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "nav2_collision_monitor/collision_detector_node.hpp"
+
+#include <exception>
+#include <utility>
+#include <functional>
+
+#include "tf2_ros/create_timer_ros.h"
+
+#include "nav2_util/node_utils.hpp"
+
+using namespace std::chrono_literals;
+
+namespace nav2_collision_monitor
+{
+
+CollisionDetector::CollisionDetector(const rclcpp::NodeOptions & options)
+: nav2_util::LifecycleNode("collision_detector", "", options)
+{
+}
+
+CollisionDetector::~CollisionDetector()
+{
+  polygons_.clear();
+  sources_.clear();
+}
+
+nav2_util::CallbackReturn
+CollisionDetector::on_configure(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Configuring");
+
+  // Transform buffer and listener initialization
+  tf_buffer_ = std::make_shared<tf2_ros::Buffer>(this->get_clock());
+  auto timer_interface = std::make_shared<tf2_ros::CreateTimerROS>(
+    this->get_node_base_interface(),
+    this->get_node_timers_interface());
+  tf_buffer_->setCreateTimerInterface(timer_interface);
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  state_pub_ = this->create_publisher<nav2_msgs::msg::CollisionDetectorState>(
+    "~/state", rclcpp::SystemDefaultsQoS());
```

Comment:
"~/state" is better to be parameterized, similarly to how it made in `collision_monitor_node.cpp`

Reply:
^ I don't necessarily think it needs to be parameterized, but it needs to be made just `state` without the `~/`

Reply:
Okay, in CollisionDetector logic, the state is mandatory parameter, opposite of CollisionMonitor logic (where this parameter is being optional), so I would agree that it could be hard-coded.
However, to be aligned with https://github.com/ros-planning/navigation2/commit/b8d077eebb7755ae3b7510ae7a993c499c8bed6c, we might need to rename this parameter to something like `collision_detector_state`, or rename back `collision_monitor_state` -> to just `state` there (second option seems better for me).

Reply:
> but it needs to be made just state without the ~/

Why though? I personally would have named both topics "~/state" for both CD and CM, so the result is "/collision_detector/state" and "/collision_monitor/state". I usually like to have the node name in the topic to know instinctively where it comes from, especially for such a generic topic like "state".
Also I usually don't parametrize topic names as they can easily be remapped if the user wants to.

Anyway no strong opinion, I'm fine with calling it "collision_detector_state" as well. Let me know

> second option seems much easier and better for me

That would break things for someone already using the CM with the default configuration, I don't think it's necessary given the other options.